sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,166 @@
+package main // import "github.com/inclavare-containers/sgx-tools"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"github.com/go-restruct/restruct"
+	"github.com/golang/protobuf/proto"
+	"github.com/inclavare-containers/rune/libenclave/intelsgx"
+	pb "github.com/inclavare-containers/sgx-tools/proto"
+	"github.com/sirupsen/logrus"
+	"net"
+)
+
+const (
+	aesmdSocket = "/var/run/aesmd/aesm.socket"
+)
+
+func dialAesmd() (*net.UnixConn, error) {
+	addr, err := net.ResolveUnixAddr("unix", aesmdSocket)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func transmitAesmd(conn *net.UnixConn, req *pb.AesmServiceRequest) ([]byte, error) {
+	rdata, err := proto.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	msgSize := uint32(len(rdata))
+	byteBuf := bytes.NewBuffer([]byte{})
+	binary.Write(byteBuf, binary.LittleEndian, &msgSize)
+	if _, err = conn.Write(byteBuf.Bytes()); err != nil {
+		return nil, err
+	}
+
+	if _, err = conn.Write(rdata); err != nil {
+		return nil, err
+	}
+
+	rdata = append(rdata[:4])
+	if _, err = conn.Read(rdata); err != nil {
+		return nil, err
+	}
+
+	byteBuf = bytes.NewBuffer(rdata)
+	if err = binary.Read(byteBuf, binary.LittleEndian, &msgSize); err != nil {
+		return nil, err
+	}
+
+	rdata = make([]byte, msgSize)
+	var msgSizeRead int
+	msgSizeRead, err = conn.Read(rdata)
+	if err != nil {
+		return nil, err
+	}
+
+	if msgSizeRead != int(msgSize) {
+		return nil, fmt.Errorf("invalid response size (returned %d, expected %d)",
+			msgSizeRead, msgSize)
+	}
+
+	return rdata, nil
+}
+
+func InitQuoteExRequest(AkId string, bPubKeyId bool, bufSize uint64) (uint64, []byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return 0, nil, err
+	}
+	defer conn.Close()
+
+	var AttkeyId []byte
+	if AkId == "" {
+		AttkeyId, err = hex.DecodeString("0000000020008c4f5775d796503e96137f77c68a829a0056ac8ded70140b081b094490c57bff000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000")
+		if err != nil {
+			return 0, nil, fmt.Errorf("Failed to decode Att key: %s", err)
+		}
+	} else {
+		AttkeyId, err = hex.DecodeString(AkId)
```

Comment:
AttKeyId must be 256*2 in string. So you can do the length check before converting it to hex string.

Reply:
> AttKeyId must be 256*2 in string. So you can do the length check before converting it to hex string.

The valid AttKey may be less than 256*2, I use the following codes to append the AttKey 

```Go
for len(AttkeyId) < 256 {
		AttkeyId = append(AttkeyId, 0)
	}
```